docs/TypeScript.md

```diff
@@ -1,291 +1,463 @@
 <h1 align="center">Fastify</h1>
 
-<a id="typescript"></a>
 ## TypeScript
-Fastify is shipped with a typings file, but it still require to install `@types/node`, depending on the Node.js version that you are using.
-
-## Types support
-We do care about the TypeScript community, but the framework is written in plain JavaScript and currently no one of the core team is a TypeScript user while only one of the collaborators is.
-We do our best to have the typing updated with the latest version of the API, but *it can happen* that the typings are not in sync.<br/>
-Luckly this is Open Source and you can contribute to fix them, we will be very happy to accept the fix and release it as soon as possible as a patch release. Checkout the [contributing](#contributing) rules!
-
-Plugins may or may not include typings. See [Plugin Types](#plugin-types) for more information.
-
-## Example
-This example TypeScript app closely aligns with the JavaScript examples:
-
-```ts
-import * as fastify from 'fastify'
-import { Server, IncomingMessage, ServerResponse } from 'http'
-
-// Create a http server. We pass the relevant typings for our http version used.
-// By passing types we get correctly typed access to the underlying http objects in routes.
-// If using http2 we'd pass <http2.Http2Server, http2.Http2ServerRequest, http2.Http2ServerResponse>
-const server: fastify.FastifyInstance<Server, IncomingMessage, ServerResponse> = fastify({})
-
-const opts: fastify.RouteShorthandOptions = {
-  schema: {
-    response: {
-      200: {
-        type: 'object',
-        properties: {
-          pong: {
-            type: 'string'
-          }
-        }
-      }
-    }
-  }
-}
 
-server.get('/ping', opts, (request, reply) => {
-  console.log(reply.res) // this is the http.ServerResponse with correct typings!
-  reply.code(200).send({ pong: 'it worked!' })
-})
-```
+The Fastify framework is written in plain JavaScript and as such type definitions are not as easy to maintain; however, since version 2 and beyond, maintainers and contributors have put in a great effort to improve the types.
 
-<a id="generic-parameters"></a>
-## Generic Parameters
-Since you can validate the querystring, params, body, and headers, you can also override the default types of those values on the request interface:
+In its current state, the type systems between Fastify v2.x and future v3.x contain breaking changes. Version 3.x type system introduces generic constraining and defaulting, plus a new way to define schema types such as a request body, querystring, and more! As the team works on improving framework and type definition synergy, sometimes parts of the API will not be typed or may be typed incorrectly. We encourage you to **contribute** to help us fill in the gaps. Just make sure to read our [`CONTRIBUTING.md`](https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md) file before getting started to make sure things go smoothly! 
```

Comment:
I think we can highlight differently the changes in v3. In the next branch, I would just refer them as the types and add a notice like so: https://github.com/fastify/fastify/blob/master/docs/Validation-and-Serialization.md#--security-notice.

Note that the website provide version-specific docs. So we can even link them directly. (for example: https://www.fastify.io/docs/v2.10.x/).

@lmammino do you think it would be possible to get a `v2.x` link as well?